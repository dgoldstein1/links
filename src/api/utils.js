import axios from "axios";
// makes request
// returns Promise{sucess, data, error}
export function makeRequest({ method, url, onErrorPrefix = "", body }) {
  return axios[`${method}`](encodeURI(url), body)
    .then(r => ({
      success: true,
      data: r.data
    }))
    .catch(e => ({
      success: false,
      error: `${onErrorPrefix}: ${e.response.data}`
    }));
}
