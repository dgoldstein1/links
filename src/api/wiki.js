import axios from "axios";

let _errOut = (s, e) => ({
  success: false,
  error: `Could not get description from wikipedia for article ${s}: ${e}`
});
/**
 * @param s {string}
 * @return JSON {
    success: bool,
    data : {
      description : string,
      img : "string URL"
    }
    error : string
  }
 **/
export function getDescription(s) {
  return _queryByTitle(s).then(r => {
    console.log(r);
    if (r.success) return Promise.resolve(r);
    // if can't find find exact title, try opensearch
    return _opensearch(s);
  });
}

// gets icon and description by trying to guess title
export function _queryByTitle(s) {
  // get extracts
  let url = process.env.REACT_APP_WIKIMETA_ENDPOINT;
  url += `?action=query`;
  url += `&format=json`;
  url += `&prop=extracts`;
  url += `&exlimit=max`;
  url += `&explaintext`;
  url += `&exintro`;
  url += `&titles=${s}`;

  return axios
    .get(encodeURI(url))
    .then(r => {
      let pageId = Object.keys(r.data.query.pages)[0];
      pageId = parseInt(pageId);
      /*jslint eqeq: true*/
      if (!pageId || pageId === -1) return _errOut(s, "no page found");
      let extract = r.data.query.pages[pageId].extract;
      // now get images
      url = process.env.REACT_APP_WIKIMETA_ENDPOINT;
      url += `?action=query`;
      url += `&prop=pageimages`;
      url += `&format=json`;
      url += `&pithumbsize=100`;
      url += `&pageids=${pageId}`;
      return axios
        .get(encodeURI(url))
        .then(r => ({
          success: true,
          description: extract,
          img: r.data.query.pages[pageId].thumbnail
        }))
        .catch(_errOut);
    })
    .catch(_errOut);
}

// gets description and image for the first
// result in an opensearch
export function _opensearch(s) {
  // get extracts
  let url = process.env.REACT_APP_WIKIMETA_ENDPOINT;
  url += "?action=opensearch";
  url += "&limit=1";
  url += "&namespace=0";
  url += "&format=json";
  url += `&search=${s}`;

  return axios
    .get(encodeURI(url))
    .then(r => {
      let pageTitle = r.data[1];
      if (!pageTitle) return _errOut(s, "no page found");
      let extract = r.data[2];
      // now get images
      url = process.env.REACT_APP_WIKIMETA_ENDPOINT;
      url += `?action=query`;
      url += `&prop=pageimages`;
      url += `&format=json`;
      url += `&pithumbsize=100`;
      url += `&titles=${pageTitle}`;
      return axios
        .get(encodeURI(url))
        .then(r => {
          let pageId = Object.keys(r.data.query.pages)[0];
          return {
            success: true,
            description: extract,
            img: pageId && r.data.query.pages[pageId].thumbnail
          };
        })
        .catch(_errOut);
    })
    .catch(_errOut);
}
