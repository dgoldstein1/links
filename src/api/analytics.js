import axios from "axios";

import { makeRequest } from "./utils";
/**
 * gets user's ip address and then posts go website analytics endpoint
 * @return: Promise
 * {
 *   success : bool,
 *   data : [{key, value}],
 *   error : string
 * }
 **/
export function postUserVisit() {
  // first get ip address of user
  let geoUrl = encodeURI(
    `/analytics/api/geoIpServer/v1?apiKey=at_Mb3nWUvk1iAL4W97H5Fs1LxAXjRCn&ipAddress=8.8.8.8`
  );
  // add in href, if exists
  return axios
    .get(geoUrl)
    .then(res => {
      if (res.data) {
        // add in referrer code
        res.data.href = new URLSearchParams(window.location.search).get("href");
        // small cleanup on attributes
        res.data.zip_code = res.data.zip;
      }
      // now post this data to the backend
      return makeRequest({
        method: "post",
        url: "/analytics/server/visits",
        onErrorPrefix: "could not post json to analytics backend",
        body: res.data
      });
    })
    .catch(e => ({
      success: false,
      error: "could not get user IP address: " + e.message
    }));
}
