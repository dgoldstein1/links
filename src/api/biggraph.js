import { makeRequest } from "./utils";
import { store } from "../reducers";
/**
 * @param {int} id
 * @return JSON {
    success: bool,
    data : []int
    error : string
  }
 **/
export function getNeighbors(id) {
  return makeRequest({
    method: "get",
    url: `${store.getState().appState.graphConfig.graphEndpoint}neighbors?node=${id}&limit=${
      store.getState().graph.maxNeighbors
    }`,
    onErrorPrefix: "Could not get neighbors"
  });
}

/**
 * gets shortest path from one node to anotther
 * @param {node} start
 * @param {node} end
 * @param {number} max number of paths to fetch
 * @param {bool} should all nodes be unique?
 * @return {promise} {success, error, data}
 **/
export function shortestPath(start, end) {
  return makeRequest({
    method: "get",
    url: `${store.getState().appState.graphConfig.graphEndpoint}shortestPath?start=${start.id}&end=${end.id}&n=${
      store.getState().graph.maxShortestPaths
    }&directed=${store.getState().graph.directedShortestPath}`,
    onErrorPrefix: `Could not get path from "${start.label}" to "${end.label}"`
  });
}

export function centrality(nodes = []) {
  return makeRequest({
    method: "post",
    url: `${store.getState().appState.graphConfig.graphEndpoint}centrality`,
    body: nodes
  });
}
