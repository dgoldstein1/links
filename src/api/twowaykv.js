import { makeRequest } from "./utils";
import { store } from "../reducers";

/**
 * fetches random nodes from twowaykv
 * @param n, number of random items to fetch
 * @return:
 *{
 *  success : bool,
 *   data : [{key, value}],
 *   error : string
 *}
 **/
export function random(n) {
  return makeRequest({
    method: "get",
    url: `${store.getState().appState.graphConfig.kvEndpoint}/random?n=${n}`,
    onErrorPrefix: "Error getting random entry"
  });
}

/**
 * gets entries from IDs
 * @param ids []int
 * @return same response type as above
 **/
export function entriesFromValues(ids) {
  return makeRequest({
    method: "post",
    url: `${
      store.getState().appState.graphConfig.kvEndpoint
    }/entriesFromValues`,
    onErrorPrefix: "Error getting entries from values",
    body: ids
  });
}

/**
 * searches 'startswith' on keys
 * @param s {string}
 * @return same response type as above
 **/
export function search(s) {
  return makeRequest({
    method: "get",
    url: `${store.getState().appState.graphConfig.kvEndpoint}/search?q=${s}`,
    onErrorPrefix: "Error searching keys"
  });
}
