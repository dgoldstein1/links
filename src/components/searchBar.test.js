import React from "react";
import SearchBar from "./searchBar";
import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import * as twowaykv from "../api/twowaykv";

configure({ adapter: new Adapter() });

describe("components", () => {
  describe("search bar", () => {
    let defaultProps = {
      onSelect: (e) => {},
      placeholder: "test",
      search: () => {},
      value: {
        label: "test1",
        key: "test-key1",
      },
    };
    it("matches snapshot", () => {
      expect(<SearchBar {...defaultProps} />).toMatchSnapshot();
    });
    describe("_onChange", () => {
      let searchArgs = undefined;
      let newEntries = ["entry1", "entry2"];
      let props = {
        ...defaultProps,
        search: (e) => {
          searchArgs = e;
          return Promise.resolve({
            success: true,
            data: {
              entries: newEntries,
            },
          });
        },
      };

      it("searches for new items", (done) => {
        searchArgs = undefined;
        let wrapper = shallow(<SearchBar {...props} />);
        expect(wrapper.state(["items"])).toEqual([]);
        wrapper.instance()._onChange({ target: { value: "entry" } });
        setTimeout(() => {
          expect(wrapper.state(["items"])).toEqual(newEntries);
          expect(searchArgs).toMatchSnapshot();
          done();
        }, 100);
      });
      it("sets new value when for on change even if bad response from kv", () => {
        let newProps = {
          ...props,
          search: (e) => {
            return Promise.resolve({
              success: false,
            });
          },
        };
        let wrapper = shallow(<SearchBar {...props} />);
        expect(wrapper.state(["value"])).toEqual(props.value.label);
        wrapper.instance()._onChange({ target: { value: "entry" } });
        setTimeout(() => {
          expect(wraooer.state(["value"])).toEqual("entry");
          done();
        }, 100);
      });
    });

    describe("_renderItem", () => {
      let wrapper = shallow(<SearchBar {...defaultProps} />);
      expect(
        wrapper.instance()._renderItem({ key: "test1" }, false)
      ).toMatchSnapshot();
    });

    describe("_onSelect", () => {
      let wrapper = shallow(<SearchBar {...defaultProps} />);
      let node = {
        id: 234,
        label: "234",
      };
      let searchItem = {
        key: "3234",
        value: 234,
      };

      expect(wrapper.state(["value"])).toEqual(defaultProps.value.label);
      wrapper.instance()._onSelect(node, searchItem);
      expect(wrapper.state(["value"])).toEqual({ id: 234, label: "234" });
    });
  });
});
