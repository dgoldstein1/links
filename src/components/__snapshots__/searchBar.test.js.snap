// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components search bar _onChange searches for new items 1`] = `"entry"`;

exports[`components search bar matches snapshot 1`] = `
<Search
  onSelect={[Function]}
  placeholder="test"
  search={[Function]}
  value={
    Object {
      "key": "test-key1",
      "label": "test1",
    }
  }
/>
`;
